File: lib/helpers/metadata.ex
defmodule Burrito.Helpers.Metadata do
  require Logger

  def run(self_path, args, release) do
    Logger.info("Generating wrapper metadata file...")

    {zig_version_string, 0} = System.cmd("zig", ["version"], cd: self_path)

    metadata_map = %{
      app_name: Atom.to_string(release.name),
      zig_version: zig_version_string |> String.trim(),
      zig_build_arguments: args,
      app_version: release.version,
      options: inspect(release.options),
      erts_version: release.erts_version |> to_string()
    }

    encoded = Jason.encode!(metadata_map)

    Path.join(self_path, "_metadata.json") |> File.write!(encoded)
  end
end

File: src/metadata.zig
const std = @import("std");

pub const MetaStruct = struct {
    app_name: []const u8 = undefined,
    zig_version: []const u8 = undefined,
    zig_build_arguments: []const []const u8 = undefined,
    app_version: []const u8 = undefined,
    options: []const u8 = undefined,
    erts_version: []const u8 = undefined,
};

pub fn parse(allocator: std.mem.Allocator, string_data: []const u8) ?MetaStruct {
    const metadata_parsed = std.json.parseFromSlice(MetaStruct, allocator, string_data, .{}) catch |e| {
        std.log.err("Error when parsing metadata: {!}", .{e});
        return null;
    };

    return metadata_parsed.value;
}
